labs/b4.md

```diff
@@ -0,0 +1,165 @@
+
+In most source tarballs, there is usually a makefile that contains a set of directives to compile a project. This is because in a project, there is usually multiple files that need to be compiled by the compiler. And even then, there are multiple compilers and compiler settings that can be used.
+
+Many projects that have to be compiled are usually in C and/or C++. On most Linux distributions, there are usually three compiler “options”. There is the GNU Toolchain which uses GCC, LLVM which uses clang, and Intel’s proprietary toolkit. Both GCC and clang are open-source and free. However, clang was created with a permissive license in response to GCC’s more viral copyleft license. Although clang was made with the intention of being completely compatible with GCC, there are some quirks and most open-source software still target GCC.
+
+To compile most software that you have the dependencies to, run:
+
+`make`
+
+This is will usually choose the correct compiler and compile the whole project. Usually when the compilation is done, the executables are in /bin or /src. 
+
+GCC and clang also have compiler flags that allow certain features to be enabled. Usually the flags that actually matter are optimization flags. Depending on what you want to optimize for, either space or memory or speed, there is a flag for it.
+
+Now, we will make a very simple application in C that prints “Hello Penguin!” named hellopenguin. Invoke:
+
+`nano hellopenguin.c`
+
+And then type in the following:
+
+```
+#include <stdio.h>
+int main()
+{
+   printf("Hello Penguin!");
+   return 0;
+}
+```
+
+Now save and exit. 
+
+We will now compile the source file that you have just written:
+
+`gcc hellopenguin.c -o hellopenguin`
+
+What this does is to take in a source file hellopenguin.c and compile it to an executable named hellopenguin with the -o output flag.
+
+# How do I package stuff?
+
+Packaging manually for Debian can be very hard and frustrating, especially for first timers. That’s why for this class, we’ll be using a really cool Ruby package called fpm which simplifies the task of packaging a lot. Please note that if you were going to package for Debian’s own repositories, this is NOT the correct/formal way to go about packaging. It is merely a great way to backport or package your own applications extremely quickly. 
+First make sure Ruby and its own package manager called Gem is installed. If they aren't, run `sudo apt install ruby-dev`. Now run the following to install fpm locally:
+
+`gem install fpm --user`
+
+Try invoking fpm and if it doesn’t work, add ~/.gem/ruby/2.3.0/bin to your PATH. To do that, add this .bashrc or just type the following into Bash:
```

Comment:
`add this .bashrc` -> `add this to your .bashrc`

Also I think it would be good to mention that running this in bash will not persist if they open a new terminal or session.